### PR TITLE
Refactor/promises

### PR DESCRIPTION
**Pull request by untra**
* refactored `ElixirSenseClient` to use promises instead of callbacks
* fixed the the [goto defintion bug](https://github.com/fr1zle/vscode-elixir/pull/75#issuecomment-318275091) , @PotterDai 
* added a file for standard promise validations across the repo

### CODE REVIEW

File: src/elixirSenseAutocompleteProvider.ts [hunk 1]
```diff
@@ -27,27 +27,23 @@ export class ElixirSenseAutocompleteProvider implements vscode.CompletionItemPro
...
-                console.log('ElixirSense client not ready');
-                console.error('rejecting');
-                reject();
-                return;
-            }
             const payload = {
                 buffer : document.getText(),
                 line   : position.line + 1,
                 column : position.character + 1
             };
```
Comment (untra) [inline src/elixirSenseAutocompleteProvider.ts]: This is how you correctly use promises, btw :1st_place_medal: 

They're kinda like elixir's pipe chains, describing the happy path chained with `.then` and with the `.catch` handling whatever gets thrown. Combined with the validator pattern, the code practically reads like english. 
It's easy to modify, and super DRY.

**Comment by untra** [comment]
@fr1zle I implore you to pull down the branch and test the plugin yourself, making sure I didn't miss anything. My desk testing has things working well, but I won't know if I missed something unless someone does a serious review of my PR.

Thanks for your assistance!

**Comment by fr1zle** [comment]
Will do, no stress. I haven't released your changes so its no big deal. And before releasing I usually try to test as much as possible. Thanks for your contribution :+1:

File: src/elixirSenseAutocompleteProvider.ts [hunk 2]
```diff
@@ -76,7 +72,7 @@ export class ElixirSenseAutocompleteProvider implements vscode.CompletionItemPro
         const matchesWordEnd = prefix.match(/\.[^A-Z][^\.]*$/);
         const matchesNonWordEnd = prefix.match(/^[^A-Z:][^\.]*$/);
         const isPrefixFunctionCall = !!(matchesWordEnd || matchesNonWordEnd);
-        if (prefix !== '' && isPrefixFunctionCall) {
+        if (prefix && !isPrefixFunctionCall) {
```
Comment (untra) [inline src/elixirSenseAutocompleteProvider.ts]: @fr1zle fixed the module dot autocomplete. I was missing a `!`
compare to the original: https://github.com/fr1zle/vscode-elixir/pull/75/files#diff-7591485bae15ac02d0a3b1808e688fa3L68

**Comment by untra** [comment]
This should be complete now. Pull down and test at your convenience. :+1:

**Comment by untra** [comment]
I'm working on one more refactor of the autocompleteprovider to support the new @impl attribute when autocompleting callbacks with elixir v1.5. If this pr can get in and approved, I can get the new autocomplete pr in.👍

Ideally, if we could get all changes in before September 1 and cut a release, that'd be super rad (ElixirConf is on September 5th!)

**Comment by fr1zle** [comment]
LGTM; TYVM :wink:

**Pull request merged**